### PR TITLE
fix: issue with streak being stuck and include new staking contracts

### PR DESCRIPTION
**Pull request by mohandast52**
## Proposed changes

- Refactored the code for the rewards history hook and component, adding necessary conditions.
- In rewards history, the most recent staking contract will now appear at the top.

Scre1

Scre31

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: frontend/constants/contractAddresses.ts [hunk 1]
```diff
@@ -16,22 +16,25 @@ export const SERVICE_REGISTRY_TOKEN_UTILITY_CONTRACT_ADDRESS: Record<
   [Chain.GNOSIS]: '0xa45E64d13A30a51b91ae0eb182e88a40e9b18eD8',
 };
 
+export const SERVICE_STAKING_CONTRACT_ADDRESSES: Record<
```
Comment (truemiller) [inline frontend/constants/contractAddresses.ts:19]: please make the const gnosis specific 🙏

File: frontend/hooks/useRewardsHistory.ts [hunk 2]
```diff
@@ -32,14 +34,15 @@ const CheckpointGraphResponseSchema = z.object({
...
   }),
 });
-type CheckpointGraphResponse = z.infer<typeof CheckpointGraphResponseSchema>;
+const CheckpointsGraphResponseSchema = z.array(CheckpointGraphResponseSchema);
+type CheckpointResponse = z.infer<typeof CheckpointGraphResponseSchema>;
 
 const fetchRewardsQuery = gql`
   {
-    checkpoints(orderBy: epoch, orderDirection: desc) {
+    checkpoints(orderBy: epoch, orderDirection: desc, first: 1000) {
```
Comment (truemiller) [inline frontend/hooks/useRewardsHistory.ts:45]: will need to consider pagination here too, the graph has 1000 limit

can also reduce the checkpoints queried too..
```gql
// only checkpoints we support in frontend
checkpoints { 
  where: { 
    contractAddress_includes: 
      [ ALL_THE_STAKING_CONTRACT_ADDRESSES as string[] ] 
  }
}
```

```gql
// no empty checkpoints
{ 
  checkpoints (where: { serviceIds_not: []}) {
    serviceIds
  }
}
```

File: frontend/constants/contractAddresses.ts (hunk 1 above)
Comment (truemiller) [reply frontend/constants/contractAddresses.ts:19]: yep exactly that :)

File: frontend/utils/service.ts [hunk 3]
```diff
@@ -37,3 +38,21 @@ export const getMinimumStakedAmountRequired = (
...
+ *
+ * Get the staking program id by address
+ * @example getStakingProgramIdByAddress('0x3052451e1eAee78e62E169AfdF6288F8791F2918') // StakingProgramId.Beta4
+ */
+export const getStakingProgramIdByAddress = (
+  contractAddress: `0x${string}`,
+): StakingProgramId | undefined => {
+  const entries = Object.entries(SERVICE_STAKING_CONTRACT_ADDRESSES) as [
+    StakingProgramId,
+    `0x${string}`,
```
Comment (Tanya-atatakai) [inline frontend/utils/service.ts:52]: tiny, can be used from here frontend/types/Address.ts

**Review by Tanya-atatakai — APPROVED** [review]

File: frontend/hooks/useRewardsHistory.ts [hunk 4]
```diff
@@ -207,54 +234,50 @@ export const useRewardsHistory = () => {
...
+    return earnedCheckpoints.reduce((streakCount, current, i) => {
+      if (isStreakBroken) return streakCount;
 
       // first iteration
       if (i === 0) {
-        const timeNow = Date.now() / 1000;
-
-        // multiplied by 2 to give a buffer of 2 days
-        const initialEpochGap = timeNow - current.epochEndTimeStamp;
+        const initialEpochGap = Math.trunc(timeNow - current.epochEndTimeStamp);
```
Comment (Tanya-atatakai) [inline frontend/hooks/useRewardsHistory.ts:253]: probably not need to truncate if we did it in timeNow

File: frontend/hooks/useRewardsHistory.ts [hunk 5]
```diff
@@ -207,54 +234,50 @@ export const useRewardsHistory = () => {
...
 
-        // if the last epoch was more than 1 day ago, break
-        if (initialEpochGap > ONE_DAY_IN_S) break;
+        // If the epoch gap is greater than the epoch length
+        if (initialEpochGap > Number(current.epochLength)) {
```
Comment (Tanya-atatakai) [inline frontend/hooks/useRewardsHistory.ts:256]: can be a few seconds greater, depending on when backend decides to close the epoch, right? unsure if it'll be somehow noticed though

**Review by Tanya-atatakai — APPROVED** [review]

File: frontend/hooks/useRewardsHistory.ts (hunk 2 above)
Comment (truemiller) [reply frontend/hooks/useRewardsHistory.ts]: i mean to process paginated data, as we're limited to 1000 records; we would need to check the next page if for example there are 1000 records in the response, and get the next 1000

**Pull request merged**